docs/Advanced-Logging.md

```diff
@@ -0,0 +1,134 @@
+
+The `Log` module also contains an optional "watch list" filter with a simple eDSL to configure it.
+When the filter is disabled, all instrumented functions will report arguments and return values.
+When enabled, invoking a watched function produces a more verbose log entry.
+
+To disable the filter, open up `hazelcore/Log.re` and set `watch_list` to `Disabled`:
+
+```reason
+let watch_list = Filter.Disabled;
+```
+
+To watch all instrumented functions, set it to `All`:
+
+``` reason
+let watch_list = Filter.All;
+```
+
+To watch a particular function, e.g., `Action_Exp.syn_perform`:
+
+```reason
+let watch_list = Filter.(md(pre("Action_Exp")) +^ fn(eq("syn_perform")));
+```
+
+Filtering works by splitting the `__FUNCTION__` string into two parts&mdash;a "module" name and a "function" name&mdash;at the last dot (".").
+
+Beware, this is a somewhat crude approximation.
+For example, tracing through an anonymous function call adds a ".(fun)" to the module name.
+So, we use string prefix equality (`pre`) on the module name (`md`) instead of whole-string equality (`eq`).
+
+As a realistic example, to trace an action-driven bug introduced by new features to existing code paths:
+
+```reason
+let watch_list =
+  Filter.(fn(has("perform")) /^ fn(has("elab")) /^ fn(has("fix_holes")));
+```
+
+and to trace a action-driven bug that only arises in expressions in synthetic position:
+
+```reason
+let watch_list =
+  Filter.(
+    md(pre("Action_Exp"))
+    /^ md(pre("Elaborator_Exp"))
+    /^ md(pre("Statics_Exp"))
+    +^ fn(pre("syn_"))
+  );
+```
+
+and for comparison, a similar one where the `fn(pre("syn_"))` constraint only applies to functions in `Action_Exp`:
+
+```reason
+let watch_list =
+  Filter.(
+    (md(pre("Action_Exp")) +^ fn(pre("syn_")))
+    /^ md(pre("Elaborator_Exp"))
+    /^ md(pre("Statics_Exp"))
+  );
+```
+
+There are two operators for composing constraints:
+
+- `+^` is conjunction (logical AND)
+- `/^` is disjunction (logical OR)
+
+and, in addition to the `eq` and `pre` string comparators, we have:
+
+- `suf` is string suffix matching
+- `has` is substring matching
+- `re` is (Perl) regular expression matching
+
+And finally, `Log` provides a few other useful functions:
+
+- `Log.debug_msg(msg)` simply writes the log level, a timestamp, and `msg`.
+- `Log.debug_state(__FUNCTION__, str, sexp)` is similar, but writes something like `"str = " ++ Sexplib.Sexp.to_string_hum(sexp)`.
+- `Log.debug_states(__FUNCTION__)` is just `debug_state` lifted to lists of (`str`, `sexp`) pairs.
+
+All three of these functions use level `Log.Debug` and none of them consult the watch list.
+They are intended for use as transient helpers and should not be committed to dev.
+
+The filtering dsl is easy to extend and will become significantly more expressive as I apply it to more complex scenarios.
+For example, in the coming weeks I'll probably add match negation, and a "call stack" to drive context-dependent selections such as:
```

Comment:
don't put a timeframe like "in the coming weeks" in dev -- just say "in the future"